### PR TITLE
Permits to delay specifying the actual indices to query in an ElasticQuery

### PR DESCRIPTION
**Pull request by andyHa**
Fixes: OX-10282

### CODE REVIEW

**Review by jmuscireum — APPROVED** [review]

File: src/main/java/sirius/db/es/Elastic.java [hunk 1]
```diff
@@ -705,34 +705,25 @@ public <E extends ElasticEntity> ElasticQuery<E> select(Class<E> type) {
...
+     * left empty and specified later via {@link ElasticQuery#withEffectiveIndices(List)}.
      *
      * @param commonSuperClass the common super class of all entities to query across
      * @param types            the types of the entities to query across
      * @param <E>              the generic common type of the entities to query across
      * @return a new query which can be used to fetch entities of the given types
      */
     @SafeVarargs
+    @SuppressWarnings("java:S1172")
+    @Explain("We only need to type for the compiler to enforce type safety.")
```
Comment (idlira) [inline src/main/java/sirius/db/es/Elastic.java:718]:
```suggestion
    @Explain("We only need the type for the compiler to enforce type safety.")
```

File: src/main/java/sirius/db/es/ElasticQuery.java [hunk 2]
```diff
@@ -328,18 +328,19 @@ private <X> List<X> autoinit(List<X> list) {
...
      * additional descriptors must share the same settings. Also note, that all entities / descriptors must share
      * the fields being queried / aggregated for this to make sense.
      *
-     * @param additionalEntitiesToQuery the additional entities to be queried
+     * @param entitiesToQuery the entities to be queried
      * @return the query itself for fluent method calls
      */
-    public ElasticQuery<E> withAdditionalIndices(Stream<Class<? extends E>> additionalEntitiesToQuery) {
-        this.additionalDescriptors = additionalEntitiesToQuery.map(type -> mixing.getDescriptor(type)).toList();
+    public ElasticQuery<E> withEffectiveIndices(List<Class<? extends E>> entitiesToQuery) {
```
Comment (idlira) [inline src/main/java/sirius/db/es/ElasticQuery.java:341]: This makes this PR technically BREAKING!

**Review by idlira — APPROVED** [review]

File: src/main/java/sirius/db/es/Elastic.java [hunk 3]
```diff
@@ -705,34 +705,25 @@ public <E extends ElasticEntity> ElasticQuery<E> select(Class<E> type) {
...
-     *
-     * @param commonSuperClass the common super class of all entities to query across
-     * @param types            the types of the entities to query across
-     * @param <E>              the generic common type of the entities to query across
-     * @return a new query which can be used to fetch entities of the given types
-     */
-    public final <E extends ElasticEntity> ElasticQuery<E> selectMultiple(Class<E> commonSuperClass,
-                                                                          List<Class<? extends E>> types) {
-        return new ElasticQuery<E>(mixing.getDescriptor(types.get(0)),
-                                   getLowLevelClient()).withAdditionalIndices(types.stream().skip(1));
```
Comment (sabieber) [inline src/main/java/sirius/db/es/Elastic.java:735]: Why? I actually hate it when APIs only allow for varargs, as one has to always convert collections to arrays to use them.

---

This is also breaking

**Review by sabieber — COMMENTED** [review]
There is no savety-net when creating an empty query and directly executing it without specifying the indices?
What is the result of such a query?

File: src/main/java/sirius/db/es/Elastic.java (hunk 3 above)
Comment (andyHa) [reply src/main/java/sirius/db/es/Elastic.java:735]: Because selectMultiple(X).withEffectiveIndices(list) is already there..

**Comment by andyHa** [comment]
> There is no savety-net when creating an empty query and directly executing it without specifying the indices? What is the result of such a query?

It was an NPE and is now an IllegalStateException...

**Review by sabieber — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**